### ISSUE TITLE
Wallet Balance Idea? (Confirmed Balance & Configurable confirmation threshold)

### ISSUE THREAD

**Issue by SoreGums** [turtlecoin/meta#124]
In an effort to clarify exactly what the state of a wallet balance is perhaps instead of "locked" and "balance" it could be reported as this

```
Confirmed Balance: X TRTL
2 Tx Pending (Y TRTL)
```

Thus whatever is in confirmed balance is spendable by the wallet.

Whatever is incoming, will become confirmed once confirmed.
The confirmation could be configurable as well, min 6, max whatever the user wants. A confirmation for clarity is block age of the tx, how many blocks deep the transaction is.

**Comment by woodyjon** [comment]
One thing to evaluate: for a user who does not know what happens in the background, I think that available and locked make sense. If he has 1000 TRTL and sends 100 TRTL. He might have an available balance of 700 TRTL and locked of 200 TRTL that he must wait to be able to spend. But if you say he has a confirmed balance of 700 TRTL and some pending transactions for 200 TRTL, he might not understand it and think that he might have sent too much.
Just thinking out loud here. Maybe a non issue.

**Comment by SoreGums** [comment]
yeah, the tricky thing with the whole cryptonote tx is that in order to do the fungible thing, you need to use mixins and you never have exact amounts that can be sent and will always need to generate "change".
the change is what gets "locked" and is technically an incoming transaction, there are different kinds of transactions or different outputs. there are 1) transactions, 2) outputs.

so while "locked" is accurate, it is also a technical term. the term is also accurate at the same time, there is an issue of clarity with regards to incoming amounts. the issue being addressed behind the scenes in this idea is how wallets represent coin balance and seeing if there is a way to make the state of the blockchain/mempool accurately portrayed.

while the suggestion in OP makes things a little bit less ideal. it does actually reflect 100% what is happening. if someone has a balance of 1,000TRTL and sent 100TRTL, with 200 in inputs resulting in 100 change, their wallet would then accurately say

```
Confirmed Balance: 700 TRTL
1 Tx Pending (100 TRTL)
```

If then at the same time another incoming transaction was also on the way for 45 TRTL then it would say 

```
Confirmed Balance: 700 TRT
2 Tx Pending (145 TRTL)
```

with the user configurable option for "confirmations" then after the tx has been committed for that number of blocks it would then be confirmed.

in summary, the whole fungibility parts of cryptonote requires further understanding over what is common knowledge, will TurtleCoin diverge at the terms for this and go with the reference of pending/confirmed? maybe, this is an idea. it is pretty simple to break down the concept of a mixin and how the fungibility part works in CN, so this idea might have merit.

The locked/balance convention while accurate isn't absolutely clear in its representation of the state of the blockchain for the given address though. so we'll see what happens :)

**Comment by SoreGums** [comment]
yeah, the tricky thing with the whole cryptonote tx is that in order to do the fungible thing, you need to use mixins and you never have exact amounts that can be sent and will always need to generate "change".
the change is what gets "locked" and is technically an incoming transaction, there aren't different kinds of transactions or different outputs. there is 1) a transaction, 2) outputs.

so while "locked" is accurate, it is also a technical term. the term is also accurate at the same time, there is an issue of clarity with regards to incoming amounts. the issue being addressed behind the scenes in this idea is how wallets represent coin balance and seeing if there is a way to make the state of the blockchain/mempool accurately portrayed.

while the suggestion in OP makes things a little bit less ideal. it does actually reflect 100% what is happening. if someone has a balance of 1,000TRTL and sent 100TRTL, with 200 in inputs resulting in 100 change, their wallet would then accurately say

```
Confirmed Balance: 800 TRTL
1 Tx Pending (100 TRTL)
```

If then at the same time another incoming transaction was also on the way for 45 TRTL then it would say 

```
Confirmed Balance: 800 TRT
2 Tx Pending (145 TRTL)
```

with the user configurable option for "confirmations" then after the tx has been committed for that number of blocks it would then be confirmed.

in summary, the whole fungibility parts of cryptonote requires further understanding over what is common knowledge, will TurtleCoin diverge at the terms for this and go with the reference of pending/confirmed? maybe, this is an idea. it is pretty simple to break down the concept of a mixin and how the fungibility part works in CN, so this idea might have merit.

The locked/balance convention while accurate isn't absolutely clear in its representation of the state of the blockchain for the given address though. so we'll see what happens :)

**Comment by brandonlehmann** [comment]
I kind of like how my bank does it...

I have a balance and an available balance. I can spend what’s available.

**Comment by SoreGums** [comment]
it's funny with the bank and visa-debit - the banks are the opposite to crypo wallets.

banks have pending withdrawals.  
crypto has pending deposits.

:P

**Comment by brandonlehmann** [comment]
If I understand what we need here, it's to added a configurable *confirmation threshold* to the walletbackend that takes into account any *unlock time/block* that would **keep the funds locked** until the confirmation threshold **and** the *unlock time/block* clears. Correct?
